velox/functions/lib/Re2Functions.cpp

```diff
@@ -383,13 +383,220 @@ class Re2SearchAndExtract final : public VectorFunction {
+  size_t fixedLength = 0;
+  for (auto i = start; i < end; i++) {
+    if (subPatternKinds[i] == SubPatternKind::kLiteralString) {
+      subPatterns.push_back(
+          {SubPatternKind::kLiteralString,
+           fixedLength,
+           subPatternRanges[i].second});
+    } else if (subPatternKinds[i] == SubPatternKind::kSingleCharWildcard) {
+      subPatterns.push_back(
+          {SubPatternKind::kSingleCharWildcard,
+           fixedLength,
+           subPatternRanges[i].second});
+    } else {
+      VELOX_UNREACHABLE();
+    }
+    fixedLength += subPatternRanges[i].second;
+  }
+
+  return fixedLength;
+}
+
+// Return the length of the fixed part(literal chars or single char wildcard) of
+// the pattern, it is mainly used to get the length of the fixed part in a
+// pattern, so we can extract the fixed part out.
+size_t fixedLength(
+    const std::vector<SubPatternKind>& subPatternKinds,
+    const std::vector<std::pair<size_t, size_t>>& subPatternRanges) {
+  size_t result = 0;
+  for (auto i = 0; i < subPatternKinds.size(); i++) {
+    if (subPatternKinds[i] != SubPatternKind::kAnyCharsWildcard) {
+      result += subPatternRanges[i].second;
+    }
+  }
+
+  return result;
+}
+
+// Get the num of bytes of current char(might be unicode char).
+size_t getCurrentCharBytes(const char* str) {
+  auto chrOffset = utf8proc_char_length(str);
+  // Skip bad byte if we get utf length < 0.
+  return UNLIKELY(chrOffset < 0) ? 1 : chrOffset;
+}
+
+} // namespace
+
 // Match string 'input' with a fixed pattern (with no wildcard characters).
 bool matchExactPattern(
     StringView input,
     const std::string& pattern,
     size_t length) {
-  return input.size() == pattern.size() &&
-      std::memcmp(input.data(), pattern.data(), length) == 0;
+  if (FOLLY_LIKELY(pattern.size() > 0)) {
+    return input.size() == pattern.size() &&
+        std::memcmp(input.data(), pattern.data(), length) == 0;
+  }
+
+  return input.size() == 0;
+}
+
+bool matchRelaxedFixedForwardAscii(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kLiteralString &&
+        std::memcmp(
+            input.data() + start + subPattern.start,
+            patternMetadata.fixedPattern().data() + subPattern.start,
+            subPattern.length) != 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool matchRelaxedFixedForwardUnicode(
+    StringView input,
+    const PatternMetadata& patternMetadata,
+    size_t start) {
+  // Compare the length first.
+  if (input.size() - start < patternMetadata.length()) {
+    return false;
+  }
+
+  auto cursor = start;
+  for (const auto& subPattern : patternMetadata.subPatterns()) {
+    if (subPattern.kind == SubPatternKind::kSingleCharWildcard) {
```

Comment:
I thought we decided to store only kLiteralString sub-patterns and compute the number of characters between these on the fly.

Reply:
In previous design, we didn't notice we need to handle UTF8, we only stored kLiteralString sub-patterns, and we only need to match kLiteralString sub-patterns, because kSingleCharWildcard are skipped -- in ASCII input, just skipping the corresponding length is enough for kSingleCharWildcard.

Now to handle UTF8, we need to not only match for kLiteralString but also for kSingleCharWildcard, due to the multibyte nature of unicode. If we don't store kSingleCharWildcard patterns in `patternMetadata.subPatterns()`, when iterate all the subPatterns, we need to calculate the info for kSingleCharWildcard pattern on the fly, the code will be complex and strange?

Reply:
I'm not sure I understand. Could you maybe illustrate this point with an example.

My thinking is that kRelaxed patter is a list of possibly unicode fixed strings separated by one or more _. Each undercore should match one character, but the number of characters to match is just the difference in bytes between the end of the one fixed sub-string and the start of the next.

Reply:
Technically you are right. But to write the code without storing the kSingleCharWildcard pattern will make the code complex, the following are pesudo code if we only store literal pattern:

```
std::vector<SubPatternMetadata> literalPatterns;

// Current matching literal sub-pattern.
int patternIndex = 0;
//  End index of previous literal pattern.
int previousLiteralPatternEnd = 0
while(patternIndex < literalPatterns.size()) {
    SubPatternMetadata subPattern = literalPatterns[patternIndex];

   // Handle the single char wildcard pattern
    if (subPattern.start > previousLiteralPatternEnd) {
       // match the kSingleCharWildcard pattern.
    }

    // match the current literal pattern
    ....

    previousLiteralPatternEnd = subPattern.start + subPattern.length;

    patternIndex++;
}
```

If we store all the sub-patterns, the code is simpler:

```
for (auto pattern : subPatterns) {
    if (pattern.kind = kLiteral) {
        // handle literal
    } else if (...) {
       // handle single wildcard...
    }
}
```